### PR TITLE
Verify token before prepending events from back-pagination

### PR DESCRIPTION
**Pull request by jplatte**
Basic fix for https://github.com/vector-im/element-x-android/issues/1076, https://github.com/vector-im/element-x-ios/issues/1787 (no cancellation of in-flight requests on timeline clear), also prepares for clearing *parts* of the timeline to constrain memory usage.

### CODE REVIEW

File: crates/matrix-sdk-ui/src/timeline/pagination.rs
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/timeline/pagination.rs:75]: nit: spurious async here? I don't see any await in this async block :eyes:

File: crates/matrix-sdk-ui/src/timeline/pagination.rs
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/timeline/pagination.rs:100]: nit: we could remove the "u16 overflow" part, it's an implementation detail that might not ring any bell to a consumer of this API

File: crates/matrix-sdk-ui/src/timeline/pagination.rs
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/timeline/pagination.rs:52]: I find it a bit weird that we're setting the back_pagination_status both here and in the caller. Instead, could we remove the status data from the `Break` type, and always set the status in this function? (except for errors, of course)

**Review by bnjbvr — APPROVED** [review]
Looks good! Considering it's quite important, approving with comments that I hope would be addressed/answered to; please re-request another review if the code changes a bunch.

Would be nice to see an extra test, see below for details. Perhaps app developers could also test this and confirm whether it works fine or not?

File: crates/matrix-sdk-ui/tests/integration/timeline/pagination.rs
Comment (jplatte) [reply crates/matrix-sdk-ui/tests/integration/timeline/pagination.rs:446]: If the timeline reset didn't happen, the amount of timeline items would be different at the end.

Comment (jplatte) [reply crates/matrix-sdk-ui/tests/integration/timeline/pagination.rs:446]: Also, there wouldn't be two pagination requests, which is also tested for.

File: crates/matrix-sdk-ui/src/timeline/pagination.rs
Comment (jplatte) [reply crates/matrix-sdk-ui/src/timeline/pagination.rs:75]: Ah yeah, this would ideally be a try block now. I guess it makes sense to change to to an immediately evaluated closure.

File: crates/matrix-sdk-ui/src/timeline/pagination.rs
Comment (jplatte) [reply crates/matrix-sdk-ui/src/timeline/pagination.rs:100]: Hmm, okay.

File: crates/matrix-sdk-ui/tests/integration/timeline/pagination.rs
Comment (bnjbvr) [reply crates/matrix-sdk-ui/tests/integration/timeline/pagination.rs:446]: It would also increase our test coverage :innocent:

Comment (jplatte) [reply crates/matrix-sdk-ui/tests/integration/timeline/pagination.rs:446]: The documentation for `subscribe_sync_gap` says that we want to remove it in the future. I'm pretty sure we can remove it now since nothing else uses it and room update handlers fill the role of the "making event handlers more general" that that doc comment mentions.

File: crates/matrix-sdk-ui/src/timeline/inner/state.rs
Comment (jplatte) [reply crates/matrix-sdk-ui/src/timeline/inner/state.rs:81]: There's two cases to consider here, a pending local echo and a failed local echo. I think the latter is more problematic. Can we postpone additional tests to after this PR is merged though? It already adds three tests and increases coverage overall.

File: crates/matrix-sdk-ui/src/timeline/pagination.rs
Comment (jplatte) [reply crates/matrix-sdk-ui/src/timeline/pagination.rs:52]: By declaring the return type to be `BackPaginationStatus`, we can't accidentally introduce new early returns that don't set the status.

File: crates/matrix-sdk-ui/src/timeline/inner/state.rs
Comment (bnjbvr) [reply crates/matrix-sdk-ui/src/timeline/inner/state.rs:81]: Since it's something that *will* backfire, I would strongly suggest to address it now, with tests at least documenting the current behavior.

**Comment by jplatte** [comment]
(rebased because this branch was a bit outdated)

**Pull request merged**
